pkg/controller/flinkapplication/flink_state_machine.go

```diff
@@ -381,6 +349,40 @@ func (s *FlinkStateMachine) handleApplicationSavepointing(ctx context.Context, a
 	return statusUnchanged, nil
 }
 
+func (s *FlinkStateMachine) handleApplicationRecovering(ctx context.Context, app *v1beta1.FlinkApplication) (bool, error) {
+	// we're in the middle of a deploy, and savepointing has failed in some way... we're going to try to recover
+	// and push through if possible
+	if rollback, reason := s.shouldRollback(ctx, app); rollback {
+		// we failed to recover, attempt to rollback
+		s.flinkController.LogEvent(ctx, app, corev1.EventTypeWarning, "JobsSubmissionFailed",
+			fmt.Sprintf("Failed to recover with externalized checkpoint: %s", reason))
+		s.updateApplicationPhase(app, v1beta1.FlinkApplicationRollingBackJob)
+		return statusChanged, nil
+	}
+
+	// TODO: the old job might still be running at this point... we should maybe try to force cancel it
+	//       (but if the JM is unavailable, our options there might be limited)
+
+	// try to find an externalized checkpoint
+	path, err := s.flinkController.FindExternalizedCheckpoint(ctx, app, app.Status.DeployHash)
```

Comment:
Should we have this behind a config or setting ?

Reply:
This isn't changing anything from the current behavior, just moving things around.